File: src/linear.py
import torch
import numpy as np
import os

# **Parameter Configuration**
BATCH_SIZE = 1      # Batch size (1 for simplicity)
IN_FEATURES = 64    # Number of input features
OUT_FEATURES = 32   # Number of output features

# **Ensure 'data' directory exists**
os.makedirs("data", exist_ok=True)

# **1. Generate input data, weights, and bias**
torch.manual_seed(42)
X = torch.rand(BATCH_SIZE, IN_FEATURES, dtype=torch.float32) * 2 - 1  # Input in range [-1, 1]
W = torch.rand(OUT_FEATURES, IN_FEATURES, dtype=torch.float32) * 2 - 1  # Weights in range [-1, 1]
B = torch.rand(OUT_FEATURES, dtype=torch.float32) * 2 - 1  # Bias in range [-1, 1]

# **2. Compute Linear transformation**
linear_layer = torch.nn.Linear(IN_FEATURES, OUT_FEATURES, bias=True)
linear_layer.weight.data = W
linear_layer.bias.data = B

Y = linear_layer(X).detach().numpy()  # Compute output

# **3. Save input, weights, bias, and output**
np.save("data/X_linear.npy", X.numpy())   # Input data
np.save("data/W_linear.npy", W.numpy())   # Weights
np.save("data/B_linear.npy", B.numpy())   # Bias
np.save("data/Y_torch_linear.npy", Y)     # PyTorch computed output

np.savetxt("data/X_linear.txt", X.numpy().flatten(), fmt="%.6f")
np.savetxt("data/W_linear.txt", W.numpy().flatten(), fmt="%.6f")
np.savetxt("data/B_linear.txt", B.numpy(), fmt="%.6f")
np.savetxt("data/Y_torch_linear.txt", Y.flatten(), fmt="%.6f")

print("✅ Linear layer data saved in `data/` directory.")

File: src/linear.cpp
#include <iostream>
#include <fstream>
#include <vector>

#define BATCH_SIZE 1
#define IN_FEATURES 64
#define OUT_FEATURES 32
typedef float fm_t;

// **Linear Layer Computation**
void linear(fm_t X[BATCH_SIZE][IN_FEATURES], fm_t W[OUT_FEATURES][IN_FEATURES], fm_t B[OUT_FEATURES], fm_t Y[BATCH_SIZE][OUT_FEATURES]) {
    for (int batch = 0; batch < BATCH_SIZE; batch++) {
        for (int out = 0; out < OUT_FEATURES; out++) {
            fm_t sum = B[out];  // Initialize with bias
            for (int in = 0; in < IN_FEATURES; in++) {
                sum += X[batch][in] * W[out][in];  // Matrix multiplication
            }
            Y[batch][out] = sum;
        }
    }
}

// **Load data from file**
void load_data(const std::string &filename, std::vector<fm_t> &data) {
    std::ifstream file(filename);
    if (!file) {
        std::cerr << "Cannot open file: " << filename << std::endl;
        exit(1);
    }
    fm_t value;
    while (file >> value) {
        data.push_back(value);
    }
}

int main() {
    // **Create Buffers**
    fm_t X[BATCH_SIZE][IN_FEATURES];
    fm_t W[OUT_FEATURES][IN_FEATURES];
    fm_t B[OUT_FEATURES];
    fm_t Y[BATCH_SIZE][OUT_FEATURES];

    // **Load Data**
    std::vector<fm_t> x_data, w_data, b_data;
    load_data("data/X_linear.txt", x_data);
    load_data("data/W_linear.txt", w_data);
    load_data("data/B_linear.txt", b_data);

    // **Fill Buffers**
    int index = 0;
    for (int batch = 0; batch < BATCH_SIZE; batch++)
        for (int in = 0; in < IN_FEATURES; in++)
            X[batch][in] = x_data[index++];

    index = 0;
    for (int out = 0; out < OUT_FEATURES; out++)
        for (int in = 0; in < IN_FEATURES; in++)
            W[out][in] = w_data[index++];

    for (int out = 0; out < OUT_FEATURES; out++)
        B[out] = b_data[out];

    // **Execute Linear Computation**
    linear(X, W, B, Y);

    // **Save Computed Output**
    std::ofstream file("data/Y_cpp_linear.txt");
    if (!file) {
        std::cerr << "Cannot write to file: data/Y_cpp_linear.txt" << std::endl;
        return 1;
    }
    for (int batch = 0; batch < BATCH_SIZE; batch++)
        for (int out = 0; out < OUT_FEATURES; out++)
            file << Y[batch][out] << "\n";

    std::cout << "✅ C++ Linear computation completed, results saved in `data/Y_cpp_linear.txt`" << std::endl;
    return 0;
}
